fabfile.py

```diff
@@ -46,258 +68,281 @@
+             " --exit-on-error --jobs=2 ~/dbdump",
+             user="postgres")
+    else:
+        sudo("psql --dbname=" + pipes.quote(env.dbname) +
+             " --set='ON_ERROR_STOP=on' -f ~/dbdump",
+             user="postgres")
 
-@task
-def recreate_encrypted_partition():
-    """ Blow away any previous encrypted partition and create a new one. """
+    # Run Django migrations, etc.
+    refresh()
 
-    create_encrypted_partition()
+    # Cleanup
+    run("rm -f ~/dbdump")
 
-@task
-def vagrant_dev_setup(dbuser=None, dbfile=None):
-    """ Set up the development environment.
-        If the dbuser and dbfile arguments are supplied, sets up
-        encrypted database storage and loads a DB dump. """
-
-    if dbuser is not None and dbfile is None:
-        raise Exception('You must provide a database dump file to load in the dbfile argument.')
-    if dbuser is None and dbfile is not None:
-        raise Exception('You must specify the PostgreSQL user that your database belongs to in the dbuser argument.')
-    using_db_dump = (dbuser is not None and dbfile is not None)
-
-    if using_db_dump:
-        load_db_dump(dbuser, dbfile)
-    else:
-        create_settings()
-        initialize_db()
-    setup_apache()
-    link_media()
+def gen_password(length):
+    return "".join([random.choice(string.letters + string.digits) for i in range(length)])
 
 @task
-def run_devserver():
-    """ Run Django dev server on port 8000. """
-    with esp_env():
-        sudo('python manage.py runserver 0.0.0.0:8000')
+def refresh():
+    """
+    Re-synchronize the remote environment with the codebase. For use when
+    switching branches or jumping around the history in git.
+
+      - upgrades and downgrades Python packages
+      - removes orphaned *.pyc files
+      - runs Django database migrations
+    """
+    ensure_environment()
+
+    with cd(env.rbase):
+        # Update the virtualenv with correct packages
+        run("pip install --upgrade -r " + env.rbase + "esp/requirements.txt")
+
+        # Clean up *.pyc files that contain stale code. In practice, the only
+        # files that cause problems are orphaned *.pyc's where the corresponding
+        # *.py file has been deleted (git leaves the *.pyc in place). Other than
+        # this, *.pyc files cannot be stale because git updates the lastmod time
+        # when changing a file, so a freshly checked out *.py file will always
+        # be newer than its *.pyc, triggering a recompile. This is an optimization
+        # over 'manage.py clean_pyc', which deletes all *pyc's in the source
+        # tree.. Note that this command is run in the Vagrant VM, since the host
+        # may not have find installed.
+        run("""find . -name '*.pyc' -exec bash -c 'test ! -f "${1%c}"' -- {} \; -delete""")
+
+        # Run Django migrations
+        manage("migrate")
+
+        # Recompile theme (run twice, to work around bug)
+        manage("recompile_theme")
```

Comment:
The specific bug this was intended to work around: after you load a database dump with a customized theme, the first recompile_theme call will fail, and the second will succeed. This won't work without the warn_only=True from the old version, or something that has a similar effect. It's not necessary when just changing code.